src/xdp/program.c

```diff
@@ -687,12 +687,12 @@ XdpParseFrame(
     Offset += sizeof(*Cache->EthHdr);
 
     if (Cache->EthHdr->Type == RtlUshortByteSwap(ETHERNET_TYPE_IPV4)) {
-        if (Buffer->DataLength < Offset + sizeof(*Cache->Ip4Hdr)) {
+        if (Buffer->DataLength < Offset + (((UINT32)Cache->Ip4Hdr->HeaderLength) << 2)) {
             goto BufferTooSmall;
         }
         Cache->Ip4Hdr = (IPV4_HEADER *)&Va[Offset];
         Cache->Ip4Valid = TRUE;
-        Offset += sizeof(*Cache->Ip4Hdr);
+        Offset += (((UINT32) Cache->Ip4Hdr->HeaderLength) << 2);
```

Comment:
We need to be careful: what if the IP header advertises a header length of 0? We could infinite loop here, right?

Reply:
LGTM. Thanks for resolving our comments!